// klaviyo-api/dist/model/campaignRecipientEstimationJobResponseObjectResourceAttributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CampaignRecipientEstimationJobResponseObjectResourceAttributes = void 0;
class CampaignRecipientEstimationJobResponseObjectResourceAttributes {
    static getAttributeTypeMap() {
        return CampaignRecipientEstimationJobResponseObjectResourceAttributes.attributeTypeMap;
    }
}
exports.CampaignRecipientEstimationJobResponseObjectResourceAttributes = CampaignRecipientEstimationJobResponseObjectResourceAttributes;
CampaignRecipientEstimationJobResponseObjectResourceAttributes.discriminator = undefined;
CampaignRecipientEstimationJobResponseObjectResourceAttributes.attributeTypeMap = [
    {
        "name": "status",
        "baseName": "status",
        "type": "CampaignRecipientEstimationJobResponseObjectResourceAttributes.StatusEnum"
    }
];
(function (CampaignRecipientEstimationJobResponseObjectResourceAttributes) {
    let StatusEnum;
    (function (StatusEnum) {
        StatusEnum[StatusEnum["Cancelled"] = 'cancelled'] = "Cancelled";
        StatusEnum[StatusEnum["Complete"] = 'complete'] = "Complete";
        StatusEnum[StatusEnum["Processing"] = 'processing'] = "Processing";
        StatusEnum[StatusEnum["Queued"] = 'queued'] = "Queued";
    })(StatusEnum = CampaignRecipientEstimationJobResponseObjectResourceAttributes.StatusEnum || (CampaignRecipientEstimationJobResponseObjectResourceAttributes.StatusEnum = {}));
})(CampaignRecipientEstimationJobResponseObjectResourceAttributes = exports.CampaignRecipientEstimationJobResponseObjectResourceAttributes || (exports.CampaignRecipientEstimationJobResponseObjectResourceAttributes = {}));
