// klaviyo-api/dist/model/catalogItemCreateQueryResourceObjectAttributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CatalogItemCreateQueryResourceObjectAttributes = void 0;
class CatalogItemCreateQueryResourceObjectAttributes {
    constructor() {
        this['integrationType'] = CatalogItemCreateQueryResourceObjectAttributes.IntegrationTypeEnum.Custom;
        this['catalogType'] = '$default';
        this['published'] = true;
    }
    static getAttributeTypeMap() {
        return CatalogItemCreateQueryResourceObjectAttributes.attributeTypeMap;
    }
}
exports.CatalogItemCreateQueryResourceObjectAttributes = CatalogItemCreateQueryResourceObjectAttributes;
CatalogItemCreateQueryResourceObjectAttributes.discriminator = undefined;
CatalogItemCreateQueryResourceObjectAttributes.attributeTypeMap = [
    {
        "name": "externalId",
        "baseName": "external_id",
        "type": "string"
    },
    {
        "name": "integrationType",
        "baseName": "integration_type",
        "type": "CatalogItemCreateQueryResourceObjectAttributes.IntegrationTypeEnum"
    },
    {
        "name": "title",
        "baseName": "title",
        "type": "string"
    },
    {
        "name": "price",
        "baseName": "price",
        "type": "number"
    },
    {
        "name": "catalogType",
        "baseName": "catalog_type",
        "type": "string"
    },
    {
        "name": "description",
        "baseName": "description",
        "type": "string"
    },
    {
        "name": "url",
        "baseName": "url",
        "type": "string"
    },
    {
        "name": "imageFullUrl",
        "baseName": "image_full_url",
        "type": "string"
    },
    {
        "name": "imageThumbnailUrl",
        "baseName": "image_thumbnail_url",
        "type": "string"
    },
    {
        "name": "images",
        "baseName": "images",
        "type": "Array<string>"
    },
    {
        "name": "customMetadata",
        "baseName": "custom_metadata",
        "type": "object"
    },
    {
        "name": "published",
        "baseName": "published",
        "type": "boolean"
    }
];
(function (CatalogItemCreateQueryResourceObjectAttributes) {
    let IntegrationTypeEnum;
    (function (IntegrationTypeEnum) {
        IntegrationTypeEnum[IntegrationTypeEnum["Custom"] = '$custom'] = "Custom";
    })(IntegrationTypeEnum = CatalogItemCreateQueryResourceObjectAttributes.IntegrationTypeEnum || (CatalogItemCreateQueryResourceObjectAttributes.IntegrationTypeEnum = {}));
})(CatalogItemCreateQueryResourceObjectAttributes = exports.CatalogItemCreateQueryResourceObjectAttributes || (exports.CatalogItemCreateQueryResourceObjectAttributes = {}));
