// klaviyo-api/dist/model/couponCodeCreateJobResponseObjectResourceAttributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CouponCodeCreateJobResponseObjectResourceAttributes = void 0;
class CouponCodeCreateJobResponseObjectResourceAttributes {
    constructor() {
        this['completedCount'] = 0;
        this['failedCount'] = 0;
    }
    static getAttributeTypeMap() {
        return CouponCodeCreateJobResponseObjectResourceAttributes.attributeTypeMap;
    }
}
exports.CouponCodeCreateJobResponseObjectResourceAttributes = CouponCodeCreateJobResponseObjectResourceAttributes;
CouponCodeCreateJobResponseObjectResourceAttributes.discriminator = undefined;
CouponCodeCreateJobResponseObjectResourceAttributes.attributeTypeMap = [
    {
        "name": "jobId",
        "baseName": "job_id",
        "type": "string"
    },
    {
        "name": "status",
        "baseName": "status",
        "type": "CouponCodeCreateJobResponseObjectResourceAttributes.StatusEnum"
    },
    {
        "name": "createdAt",
        "baseName": "created_at",
        "type": "Date"
    },
    {
        "name": "totalCount",
        "baseName": "total_count",
        "type": "number"
    },
    {
        "name": "completedCount",
        "baseName": "completed_count",
        "type": "number"
    },
    {
        "name": "failedCount",
        "baseName": "failed_count",
        "type": "number"
    },
    {
        "name": "completedAt",
        "baseName": "completed_at",
        "type": "Date"
    },
    {
        "name": "errors",
        "baseName": "errors",
        "type": "Array<APIJobErrorPayload>"
    },
    {
        "name": "expiresAt",
        "baseName": "expires_at",
        "type": "Date"
    }
];
(function (CouponCodeCreateJobResponseObjectResourceAttributes) {
    let StatusEnum;
    (function (StatusEnum) {
        StatusEnum[StatusEnum["Cancelled"] = 'cancelled'] = "Cancelled";
        StatusEnum[StatusEnum["Complete"] = 'complete'] = "Complete";
        StatusEnum[StatusEnum["Processing"] = 'processing'] = "Processing";
        StatusEnum[StatusEnum["Queued"] = 'queued'] = "Queued";
    })(StatusEnum = CouponCodeCreateJobResponseObjectResourceAttributes.StatusEnum || (CouponCodeCreateJobResponseObjectResourceAttributes.StatusEnum = {}));
})(CouponCodeCreateJobResponseObjectResourceAttributes = exports.CouponCodeCreateJobResponseObjectResourceAttributes || (exports.CouponCodeCreateJobResponseObjectResourceAttributes = {}));
