// klaviyo-api/dist/model/couponCodeUpdateQueryResourceObjectAttributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CouponCodeUpdateQueryResourceObjectAttributes = void 0;
class CouponCodeUpdateQueryResourceObjectAttributes {
    static getAttributeTypeMap() {
        return CouponCodeUpdateQueryResourceObjectAttributes.attributeTypeMap;
    }
}
exports.CouponCodeUpdateQueryResourceObjectAttributes = CouponCodeUpdateQueryResourceObjectAttributes;
CouponCodeUpdateQueryResourceObjectAttributes.discriminator = undefined;
CouponCodeUpdateQueryResourceObjectAttributes.attributeTypeMap = [
    {
        "name": "status",
        "baseName": "status",
        "type": "CouponCodeUpdateQueryResourceObjectAttributes.StatusEnum"
    },
    {
        "name": "expiresAt",
        "baseName": "expires_at",
        "type": "Date"
    }
];
(function (CouponCodeUpdateQueryResourceObjectAttributes) {
    let StatusEnum;
    (function (StatusEnum) {
        StatusEnum[StatusEnum["AssignedToProfile"] = 'ASSIGNED_TO_PROFILE'] = "AssignedToProfile";
        StatusEnum[StatusEnum["Deleting"] = 'DELETING'] = "Deleting";
        StatusEnum[StatusEnum["Processing"] = 'PROCESSING'] = "Processing";
        StatusEnum[StatusEnum["Unassigned"] = 'UNASSIGNED'] = "Unassigned";
        StatusEnum[StatusEnum["VersionNotActive"] = 'VERSION_NOT_ACTIVE'] = "VersionNotActive";
    })(StatusEnum = CouponCodeUpdateQueryResourceObjectAttributes.StatusEnum || (CouponCodeUpdateQueryResourceObjectAttributes.StatusEnum = {}));
})(CouponCodeUpdateQueryResourceObjectAttributes = exports.CouponCodeUpdateQueryResourceObjectAttributes || (exports.CouponCodeUpdateQueryResourceObjectAttributes = {}));
