// klaviyo-api/dist/model/renderOptionsSubObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderOptionsSubObject = void 0;
class RenderOptionsSubObject {
    constructor() {
        this['shortenLinks'] = true;
        this['addOrgPrefix'] = true;
        this['addInfoLink'] = true;
        this['addOptOutLanguage'] = false;
    }
    static getAttributeTypeMap() {
        return RenderOptionsSubObject.attributeTypeMap;
    }
}
exports.RenderOptionsSubObject = RenderOptionsSubObject;
RenderOptionsSubObject.discriminator = undefined;
RenderOptionsSubObject.attributeTypeMap = [
    {
        "name": "shortenLinks",
        "baseName": "shorten_links",
        "type": "boolean"
    },
    {
        "name": "addOrgPrefix",
        "baseName": "add_org_prefix",
        "type": "boolean"
    },
    {
        "name": "addInfoLink",
        "baseName": "add_info_link",
        "type": "boolean"
    },
    {
        "name": "addOptOutLanguage",
        "baseName": "add_opt_out_language",
        "type": "boolean"
    }
];
